networkx/algorithms/approximation/steinertree.py

```diff
@@ -248,3 +249,179 @@ def steiner_tree(G, terminal_nodes, weight="weight", method=None):
+    q = deque([(source, 0)])
+    while q:
+        node, depth = q.popleft()
+        if depth == cutoff:
+            continue
+        for nbr in G[node]:
+            if nbr not in seen:
+                seen.add(nbr)
+                q.append((nbr, depth + 1))
+    return seen
+
+
+def _directed_steiner_tree_density(G, terminals, weight):
+    """Compute density = total cost / #covered terminals."""
+    if G.number_of_edges() == 0:
+        return float("inf")
+
+    terminals = set(terminals)
+    num_terminals = 0
+    total = 0
+    for u, v, d in G.edges(data=True):
+        if u in terminals:
+            terminals.remove(u)
+            num_terminals += 1
+        if v in terminals:
+            terminals.remove(v)
+            num_terminals += 1
+        total += d.get(weight, 1)
+
+    return float("inf") if num_terminals == 0 else total / num_terminals
+
+
+def _directed_steiner_tree(G, root, terminals, min_terminals, cutoff, weight):
+    """Recursive helper for directed_steiner_tree."""
+    H = nx.DiGraph()
+    if min_terminals <= 0 or not terminals or cutoff == 0:
+        return H
+
+    terminals = set(terminals)
+    if cutoff == 1:
+        H.add_node(root, **G.nodes[root])
+        edges_filtered = [
+            (u, v, d) for u, v, d in G.edges(data=True) if u == root and v in terminals
+        ]
+
+        for u, v, d in edges_filtered:
+            H.add_node(v, **G.nodes[v])
+            H.add_edge(u, v, **d)
+
+        return H
+
+    reached_terminals = set()
+    while len(reached_terminals) < min_terminals and terminals:
+        min_sub_tree = nx.DiGraph()
+        min_density = float("inf")
+
+        for v in G.successors(root):
+            for n in range(1, len(terminals) + 1):
+                sub_tree = _directed_steiner_tree(
+                    G, v, terminals, n, cutoff - 1, weight
+                )
+                sub_tree.add_node(root, **G.nodes[root])
+                sub_tree.add_node(v, **G.nodes[v])
+                sub_tree.add_edge(root, v, **G[root][v])
+
+                sub_density = _directed_steiner_tree_density(
+                    sub_tree, terminals, weight
+                )
+                if sub_density < min_density:
+                    min_sub_tree = sub_tree
+                    min_density = sub_density
+        covered_terminals = {n for n in min_sub_tree.nodes if n in terminals}
+        if not covered_terminals:
+            return H
+        reached_terminals |= covered_terminals
+        terminals -= reached_terminals
+
+        H.add_nodes_from(min_sub_tree.nodes(data=True))
+        H.add_edges_from(min_sub_tree.edges(data=True))
+
+    return H
+
+
+def directed_steiner_tree(
+    G, root, terminals, min_terminals=None, cutoff=None, weight="weight"
+):
+    """Approximation algorithm for the Directed Steiner Tree problem.
+
+    Parameters
+    ----------
+    G : DiGraph
+        A directed graph.
+    root : node
+        The root node.
+    terminals : set
```

Comment:
Since terminals can be any iterable the way you've got the conversions to `set`, this can be declared as an `iterable of nodes` rather than a set.  right?